### PR TITLE
[TARGET] Cleanup the target_host usage to new target style.

### PR DESCRIPTION
**Pull request by tqchen**
As we move towards the new Target system, most of the target specifications in the compiler already moved to the new style. This PR aims to cleanup the rest in the codebase.

- Add deprecation warnings to functions with target_host parameters.
- Update the build usage to new target style.

### CODE REVIEW

**Comment by Mousius** [comment]
Hi @tqchen,

Is there an RFC or tracking issue for this you can link to this?

**Comment by junrushao1994** [comment]
@Mousius The RFC is here: https://discuss.tvm.apache.org/t/rfc-tvm-target-specification/6844

**Comment by junrushao1994** [comment]
CC: @zxybazh

**Review by junrushao1994 — APPROVED** [review]

**Comment by tqchen** [comment]
These were the left over items that should have been done as part of the Target system refactor.

**Review by Hzfengsy — APPROVED** [review]

File: python/tvm/autotvm/task/relay_integration.py [hunk 1]
```diff
@@ -77,6 +78,11 @@ def extract_from_program(mod, params, target, target_host=None, ops=None):
     task: Array of autotvm.task.Task
         collected tasks
     """
+    if target_host is not None:
+        warnings.warn(
+            "target_host parameter is going to be deprecated. "
+            "Please pass in tvm.target.Target(target, host=target_host) instead."
+        )
```
Comment (zxybazh) [inline python/tvm/autotvm/task/relay_integration.py:85]: Shall we also add the warning to `extract_from_multiple_program`?

File: python/tvm/relay/build_module.py [hunk 2]
```diff
@@ -332,16 +337,23 @@ def build(ir_mod, target=None, target_host=None, params=None, mod_name="default"
...
+
+    target, target_host = Target.check_and_update_host_consist(
+        target, target_host, target_is_dict_key=False
+    )
+
     target = build_target_by_device_type_map(target)
     if isinstance(target_host, (str, Target)):
         target_host = Target(target_host)
     elif target_host:
         raise ValueError("target host must be the type of str, " + "tvm.target.Target, or None")
```
Comment (zxybazh) [inline python/tvm/relay/build_module.py:355]: This part seems to be obsolete after the change.

**Review by zxybazh — COMMENTED** [review]
Generally the PR looks good to me and thanks TQ for the very detailed follow up and adding warnings for target host deprecation. It's long been my hope that we can completely get rid of target host as an addtional argument and have a concise and united target representation. I am very supportive of the change and it takes a lot of effort to change all occurences of target host.

It's very easy to overlook some function calls to give warnings so here I propose to update the `check_and_update_host_consist` function as follows to check if we are dispatching a new target host for any target with no current target host.

```
    @staticmethod
    def check_and_update_host_consist(target, host=None, target_is_dict_key=True):
        """A helper function that merges a legacy "target, target_host" pair, then returns
        the merged target and its host field. The function is for legacy target and target
        host pair only, and should not be used in the new target system.

        Parameters
        ----------
        target : Union[str, Dict[str, Any], Target]
            The target or heterogeneous target
        host : Union[str, Dict[str, Any], Target, None]
            The target host
        target_is_dict_key : Bool
            When the type of target is dict, whether Target is the key (Otherwise the value)
        """
        if target is None:
            assert host is None, "Target host is not empty when target is empty."
            return target, host
        warning_flag = False
        if isinstance(target, dict) and "kind" not in target:
            new_target = {}
            for tgt, mod in target.items():
                if not target_is_dict_key:
                    tgt, mod = mod, tgt
                if isinstance(tgt, (dict, str, Target)):
                    tgt, host = Target.check_and_update_host_consist(tgt, host)
                if not target_is_dict_key:
                    tgt, mod = mod, tgt
                new_target[tgt] = mod
            target = new_target

        else:
            warning_flag = (host is not None) and (target.host is None)
            target = Target(target, host)
            host = target.host
        if warning_flag:
            warnings.warn(
                "target_host parameter is going to be deprecated. "
                "Please pass in tvm.target.Target(target, host=target_host) instead."
            )
        return target, host
```

**Comment by tqchen** [comment]
Thanks @zxybazh , given `check_and_update_host_consist` can be called multiple times, it may not be desirable to do so(for the second time it is called). So I think we can still take the explicit approach. The suggestion of `extract_from_multiple_program` is great, can you send a followup PR?

File: python/tvm/relay/build_module.py (hunk 2 above)
Comment (tqchen) [reply python/tvm/relay/build_module.py:355]: since we only issue an warning, we might still want to keep this one

**Comment by zxybazh** [comment]
Sounds good. I'll send a follow up PR.

File: python/tvm/relay/build_module.py (hunk 2 above)
Comment (zxybazh) [reply python/tvm/relay/build_module.py:355]: The type is already checked when doing `check_and_update_host_consist`, see [here](https://github.com/apache/tvm/blob/be03d62e5b0afd607964365bc73e94f72fdfaaef/python/tvm/target/target.py#L108). I'll remove it and check if there are other obsolete parts in my follow up PR.

**Pull request merged**